cmd/disk-cache.go

```diff
@@ -0,0 +1,490 @@
+	if err != nil {
+		return err
+	}
+	pipeReader.Close()
+	return
+}
+
+// Uses cached-object to serve the request. If object is not cached it serves the request from the backend and also
+// stores it in the cache for serving subsequent requests.
+func (c cacheObjects) GetObject(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error) {
+	return c.getObject(bucket, object, startOffset, length, writer)
+}
+
+// Common function for GetObjectInfo
+func (c cacheObjects) getObjectInfo(bucket, object string) (ObjectInfo, error) {
+	getObjectInfoFn := c.GetObjectInfoFn
+
+	dcache, err := c.getCache(bucket, object)
+	if err != nil {
+		return getObjectInfoFn(bucket, object)
+	}
+	objInfo, err := getObjectInfoFn(bucket, object)
+	if err != nil {
+		if _, ok := errorCause(err).(BackendDown); !ok {
+			if _, ok = errorCause(err).(ObjectNotFound); ok {
+				// Delete the cached entry if backend object was deleted.
+				dcache.Delete(bucket, object)
+			}
+			return ObjectInfo{}, err
+		}
+		return ObjectInfo{}, BackendDown{}
+	}
+	cachedObjInfo, err := dcache.GetObjectInfo(bucket, object)
+	if err != nil {
+		return objInfo, nil
+	}
+	if cachedObjInfo.ETag != objInfo.ETag {
+		// Delete the cached entry if the backend object was replaced.
+		dcache.Delete(bucket, object)
+	}
+	return objInfo, nil
+}
+
+// Returns ObjectInfo from cache or the backend.
+func (c cacheObjects) GetObjectInfo(bucket, object string) (ObjectInfo, error) {
+	return c.getObjectInfo(bucket, object)
+}
+
+// Common code for PutObject and PutObjectInfo
+func (c cacheObjects) putObject(bucket, object string, r *hash.Reader, metadata map[string]string) (objInfo ObjectInfo, err error) {
+	putObjectFn := c.PutObjectFn
+
+	dcache, err := c.getCache(bucket, object)
+	if err != nil {
+		// no cache policy for this bucket,object pair,execute backend call.
+		return putObjectFn(bucket, object, r, metadata)
+	}
+	size := r.Size()
+	objInfo = ObjectInfo{}
+	buf := make([]byte, 0, 64*1024)
+	// Initialize pipe to stream data to backend
+	pipeReader, pipeWriter := io.Pipe()
+	hashReader, err := hash.NewReader(pipeReader, size, r.MD5HexString(), r.SHA256HexString())
+	if err != nil {
+		return ObjectInfo{}, err
+	}
+	// Initialize pipe to stream data to cache
+	rPipe, wPipe := io.Pipe()
+	cHashReader, err := hash.NewReader(rPipe, size, r.MD5HexString(), r.SHA256HexString())
+	if err != nil {
+		return ObjectInfo{}, err
+	}
+	go func() {
+		objInfo, err = putObjectFn(bucket, object, hashReader, metadata)
+		if err != nil {
+			pipeWriter.CloseWithError(err)
+			return
+		}
+	}()
+	go func() {
+		if err = dcache.Put(bucket, object, cHashReader, metadata, size); err != nil {
+			wPipe.CloseWithError(err)
+			return
+		}
+	}()
+
+	for {
+		n, rerr := r.Read(buf[:cap(buf)])
+		buf = buf[:n]
+
+		if rerr != nil && rerr != io.EOF {
+			pipeWriter.CloseWithError(rerr)
+			wPipe.CloseWithError(rerr)
+			break
+		}
+		if n > 0 {
+			pipeWriter.Write(buf)
```

Comment:
Can we not simplify this with using io.MultiWriter(pipeWriter, wPipe) and then do `io.Copy(w, r)` ?